pkg/catalogv2/git/download.go

```diff
@@ -1,100 +1,111 @@
+		return "", fmt.Errorf("head failure: %w", err)
+	}
+
+	commit, err := git.currentCommit()
+	if err != nil {
+		return "", fmt.Errorf("head failure: %w", err)
 	}
 
-	return git.Head(branch)
+	return commit, nil
 }
 
+// Update updates git repo if remote sha has changed
 func Update(secret *corev1.Secret, namespace, name, gitURL, branch string, insecureSkipTLS bool, caBundle []byte) (string, error) {
 	git, err := gitForRepo(secret, namespace, name, gitURL, insecureSkipTLS, caBundle)
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("update failure: %w", err)
 	}
 
 	if isBundled(git) && settings.SystemCatalog.Get() == "bundled" {
 		return Head(secret, namespace, name, gitURL, branch, insecureSkipTLS, caBundle)
 	}
 
-	commit, err := git.Update(branch)
-	if err != nil && isBundled(git) {
-		return Head(secret, namespace, name, gitURL, branch, insecureSkipTLS, caBundle)
+	if err := git.clone(branch); err != nil {
+		return "", nil
 	}
-	return commit, err
-}
 
-func Ensure(secret *corev1.Secret, namespace, name, gitURL, commit string, insecureSkipTLS bool, caBundle []byte) error {
-	if commit == "" {
-		return nil
+	if err := git.reset("HEAD"); err != nil {
+		return "", fmt.Errorf("update failure: %w", err)
 	}
-	git, err := gitForRepo(secret, namespace, name, gitURL, insecureSkipTLS, caBundle)
+
+	commit, err := git.currentCommit()
 	if err != nil {
-		return err
+		return commit, fmt.Errorf("update failure: %w", err)
 	}
 
-	// If the repositories are rancher managed and if bundled is set
-	// don't fetch anything from upstream.
-	if isBundled(git) && settings.SystemCatalog.Get() == "bundled" {
-		return nil
+	changed, err := git.remoteSHAChanged(branch, commit)
+	if err != nil {
+		return commit, fmt.Errorf("update failure: %w", err)
+	}
+	if !changed {
+		return commit, nil
 	}
 
-	return git.Ensure(commit)
-}
+	if err := git.fetchAndReset(branch); err != nil {
+		return "", fmt.Errorf("update failure: %w", err)
+	}
 
-func isBundled(git *git) bool {
-	return strings.HasPrefix(git.Directory, staticDir) || strings.HasPrefix(git.Directory, localDir)
+	lastCommit, err := git.currentCommit()
+	if err != nil && isBundled(git) {
+		return Head(secret, namespace, name, gitURL, branch, insecureSkipTLS, caBundle)
+	}
+	return lastCommit, nil
 }
 
 func gitForRepo(secret *corev1.Secret, namespace, name, gitURL string, insecureSkipTLS bool, caBundle []byte) (*git, error) {
-	isGitSSH, err := isGitSSH(gitURL)
+	err := validateGitURL(gitURL)
 	if err != nil {
-		return nil, fmt.Errorf("failed to verify the type of URL %s: %w", gitURL, err)
-	}
-	if !isGitSSH {
-		u, err := url.Parse(gitURL)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse URL %s: %w", gitURL, err)
-		}
-		if u.Scheme != "http" && u.Scheme != "https" {
-			return nil, fmt.Errorf("invalid git URL scheme %s, only http(s) and git supported", u.Scheme)
-		}
+		return nil, fmt.Errorf("%w: only http(s) or ssh supported", err)
```

Comment:
Can we change this to `%w: only http(s) or ssh:// supported`

Reply:
OK